File: research/benincasa/q_d_census.py
"""Authoritative finite Q_D census used in the soft-axis comparison.

This materializes the previously inline audit without changing its bases,
deck-orbit completion, projection, or cutoff admission rule.
"""

from fractions import Fraction as Q
import importlib.util
import json
from pathlib import Path

P = 2305843009213693951
SECTORS = ((1, 1), (1, 0), (0, 1), (0, 0))
EXPECTED = {
    12: (105, 68, 31),
    16: (155, 106, 57),
    20: (213, 152, 91),
    24: (279, 206, 133),
}
EXPECTED_KOSZUL_AUDIT = {
    12: (16, (15, 22), (5, 20, 6)),
    16: (22, (21, 32), (7, 42, 8)),
    20: (28, (27, 42), (9, 72, 10)),
    24: (34, (33, 52), (11, 110, 12)),
}

dependency = Path(__file__).parents[1] / "voevodsky" / "check_soft_axis_deck_orbit_completion.py"
spec = importlib.util.spec_from_file_location("deck", dependency)
deck = importlib.util.module_from_spec(spec)
spec.loader.exec_module(deck)


def finite(x):
    return x.numerator * pow(x.denominator, P - 2, P) % P


def rank(columns):
    basis = {}
    for source in columns:
        vector = {row: finite(value) for row, value in source.items() if value}
        while vector:
            pivot = min(vector)
            if pivot not in basis:
                inverse = pow(vector[pivot], P - 2, P)
                basis[pivot] = {row: value * inverse % P for row, value in vector.items()}
                break
            factor = vector[pivot]
            for row, value in basis[pivot].items():
                vector[row] = (vector.get(row, 0) - factor * value) % P
                if not vector[row]:
                    vector.pop(row, None)
    return len(basis)


def component_rows(cutoff, u_degree):
    # e_a is odd under rho(a)=-a; e_b and e_u are even.
    rows = []
    for component, frame_character in ((0, -1), (1, 1), (2, 1)):
        for ud in range(u_degree + 1):
            for total in range(cutoff + 1):
                for ad in range(total + 1):
                    if (-1) ** ad * frame_character == 1:
                        rows.append((component, ud, ad, total - ad))
    return rows


def census(cutoff):
    rows = component_rows(cutoff, 1)
    position = {monomial: index for index, monomial in enumerate(rows)}
    columns = []

    def emit(parts, target=columns):
        support = [monomial for part in parts for monomial in part]
        if not support:
            return
        # Whole-column admission: never truncate individual components.
        if any(ud >= 2 or ad + bd > cutoff for ud, ad, bd in support):
            return
        column = {}
        for component, part in enumerate(parts):
            for (ud, ad, bd), coefficient in part.items():
                key = (component, ud, ad, bd)
                if key in position:  # plus-character projection
                    row = position[key]
                    column[row] = column.get(row, Q(0)) + coefficient
        column = {row: value for row, value in column.items() if value}
        if column:
            target.append(column)

    euler = (
        deck.scale(deck.a, Q(1, 4)),
        {},
        deck.scale(deck.u, Q(1, 2)),
    )

    for sa, sb in SECTORS:
        ea, eb = 2 - sa, 2 - sb
        for conjugate in (False, True):
            l2 = deck.L2_plus if conjugate else deck.L2_minus
            base = deck.mul(deck.power(deck.L1, ea), deck.power(l2, eb))
            for total in range(cutoff + 1):
                for ad in range(total + 1):
                    f = deck.mul(deck.power(deck.a, ad), deck.power(deck.b, total - ad))
                    m = deck.mul(f, base)

                    c_p = deck.scale(deck.mul(deck.derivative(f, 2), base), -1)
                    if sa:
                        c_p = deck.add(
                            c_p,
                            deck.scale(
                                deck.mul(f, deck.power(deck.L1, ea - 1), deck.power(l2, eb)),
                                sa,
                            ),
                        )
                    h_p = ({}, deck.scale(m, Q(3, 2)), {})
                    emit([deck.add(h_p[i], deck.mul(c_p, euler[i])) for i in range(3)])

                    c_q = deck.mul(deck.derivative(f, 1), base)
                    if sb:
                        c_q = deck.add(
                            c_q,
                            deck.scale(
                                deck.mul(f, deck.power(deck.L1, ea), deck.power(l2, eb - 1)),
                                -sb,
                            ),
                        )
                    h_q = (deck.scale(m, Q(-3, 2)), {}, {})
                    emit([deck.add(h_q[i], deck.mul(c_q, euler[i])) for i in range(3)])

    # Principal Euler columns E(P_D).
    for total in range(cutoff + 1):
        for ad in range(total + 1):
            p = deck.mul(deck.power(deck.a, ad), deck.power(deck.b, total - ad))
            emit([deck.mul(p, component) for component in euler])

    full_dimension = len(rows) - rank(columns)
    rows_zero = [row for row in rows if row[1] == 0]
    zero_position = {row: index for index, row in enumerate(rows_zero)}
    zero_columns = []
    for column in columns:
        restricted = {
            zero_position[rows[row]]: value
            for row, value in column.items()
            if rows[row][1] == 0
        }
        if restricted:
            zero_columns.append(restricted)
    special_dimension = len(rows_zero) - rank(zero_columns)
    torsion_dimension = 2 * special_dimension - full_dimension

    # Compute the intrinsic Koszul filtration on H(Q_D,u).  Intersections are
    # taken by ranks; in particular the relation among the three elementary
    # syzygies is retained rather than assigning monomial families by hand.
    u_columns = []
    for row, (component, ud, ad, bd) in enumerate(rows):
        if ud == 0:
            shifted = position.get((component, 1, ad, bd))
            if shifted is not None:
                u_columns.append({shifted: Q(1)})

    def frozen(poly):
        return {monomial: value for monomial, value in poly.items() if monomial[0] == 0}

    ka = frozen(deck.derivative(deck.K, 1))
    kb = frozen(deck.derivative(deck.K, 2))
    ku = frozen(deck.derivative(deck.K, 0))
    s0_columns = []
    sau_columns = []
    for total in range(cutoff + 1):
        for ad in range(total + 1):
            multiplier = deck.mul(deck.power(deck.a, ad), deck.power(deck.b, total - ad))
            emit(
                [deck.scale(deck.mul(multiplier, kb), -1), deck.mul(multiplier, ka), {}],
                s0_columns,
            )
            emit(
                [{}, deck.scale(deck.mul(multiplier, ku), -1), deck.mul(multiplier, kb)],
                s0_columns,
            )
            emit(
                [deck.scale(deck.mul(multiplier, ku), -1), {}, deck.mul(multiplier, ka)],
                sau_columns,
            )

    def multiply_column_by_u(column):
        shifted = {}
        for row, value in column.items():
            component, ud, ad, bd = rows[row]
            if ud == 0:
                target = position[(component, 1, ad, bd)]
                shifted[target] = shifted.get(target, Q(0)) + value
        return {row: value for row, value in shifted.items() if value}

    s1_columns = s0_columns + sau_columns
    rank_i = rank(columns)
    module_defect = (
        rank(columns + [multiply_column_by_u(column) for column in columns]) - rank_i
    )
    cycle_defects = (
        rank(columns + [multiply_column_by_u(column) for column in s0_columns]) - rank_i,
        rank(columns + [multiply_column_by_u(column) for column in s1_columns]) - rank_i,
    )

    b_columns = columns + u_columns
    rank_b = rank(b_columns)
    rank_b_s0 = rank(b_columns + s0_columns)
    rank_b_s1 = rank(b_columns + s1_columns)
    filtration = (
        rank_b_s0 - rank_b,
        rank_b_s1 - rank_b_s0,
        torsion_dimension - (rank_b_s1 - rank_b),
    )
    return (
        full_dimension,
        special_dimension,
        torsion_dimension,
        module_defect,
        cycle_defects,
        filtration,
    )


def main():
    results = {cutoff: census(cutoff) for cutoff in EXPECTED}
    assert {cutoff: result[:3] for cutoff, result in results.items()} == EXPECTED
    assert {cutoff: result[3:] for cutoff, result in results.items()} == EXPECTED_KOSZUL_AUDIT
    for cutoff, result in results.items():
        print(
            f"D={cutoff}: dim_Q={result[0]} dim_Q_mod_u={result[1]} t={result[2]} "
            f"module_defect={result[3]} cycle_defects={result[4]} "
            f"formal_F0={result[5][0]} formal_F1/F0={result[5][1]} formal_H/F1={result[5][2]}"
        )
    print(json.dumps({
        "schema": "marici.benincasa.q_d_census.v1",
        "results": {str(k): [*v[:4], list(v[4]), list(v[5])] for k, v in results.items()},
        "closed_form": "t_D=(D/2)^2-D/2+1",
    }, sort_keys=True))


if __name__ == "__main__":
    main()

File: research/benincasa/marici-gm/src/bin/q_d_census.rs
use std::collections::{BTreeMap, HashMap};

const P: i64 = 2_305_843_009_213_693_951;
type Mon = (u8, usize, usize);
type Poly = BTreeMap<Mon, i64>;
type Row = (usize, u8, usize, usize);
type Column = BTreeMap<usize, i64>;

fn add_mod(x: i64, y: i64) -> i64 {
    ((x as i128 + y as i128).rem_euclid(P as i128)) as i64
}
fn mul_mod(x: i64, y: i64) -> i64 {
    ((x as i128 * y as i128).rem_euclid(P as i128)) as i64
}
fn pow_mod(mut x: i64, mut n: i64) -> i64 {
    let mut out = 1;
    while n > 0 {
        if n & 1 == 1 { out = mul_mod(out, x); }
        x = mul_mod(x, x);
        n >>= 1;
    }
    out
}
fn inv(x: i64) -> i64 { pow_mod(x, P - 2) }
fn mon(u: u8, a: usize, b: usize, c: i64) -> Poly {
    BTreeMap::from([((u, a, b), c.rem_euclid(P))])
}
fn scale(x: &Poly, c: i64) -> Poly {
    x.iter().filter_map(|(&m, &v)| {
        let z = mul_mod(v, c);
        (z != 0).then_some((m, z))
    }).collect()
}
fn add(xs: &[Poly]) -> Poly {
    let mut out = Poly::new();
    for x in xs {
        for (&m, &v) in x {
            let z = add_mod(*out.get(&m).unwrap_or(&0), v);
            if z == 0 { out.remove(&m); } else { out.insert(m, z); }
        }
    }
    out
}
fn mul(x: &Poly, y: &Poly) -> Poly {
    let mut out = Poly::new();
    for (&(u1, a1, b1), &c1) in x {
        for (&(u2, a2, b2), &c2) in y {
            if u1 + u2 >= 2 { continue; }
            let m = (u1 + u2, a1 + a2, b1 + b2);
            let z = add_mod(*out.get(&m).unwrap_or(&0), mul_mod(c1, c2));
            if z == 0 { out.remove(&m); } else { out.insert(m, z); }
        }
    }
    out
}
fn power(x: &Poly, n: usize) -> Poly {
    let mut out = mon(0, 0, 0, 1);
    for _ in 0..n { out = mul(&out, x); }
    out
}
fn derivative(x: &Poly, coordinate: usize) -> Poly {
    let mut out = Poly::new();
    for (&(u, a, b), &c) in x {
        let degrees = [u as usize, a, b];
        if degrees[coordinate] == 0 { continue; }
        let mut target = degrees;
        target[coordinate] -= 1;
        out.insert(
            (target[0] as u8, target[1], target[2]),
            mul_mod(c, degrees[coordinate] as i64),
        );
    }
    out
}

fn rank(columns: Vec<Column>) -> usize {
    let mut basis: BTreeMap<usize, Column> = BTreeMap::new();
    for mut vector in columns {
        while let Some((&pivot, &value)) = vector.first_key_value() {
            if let Some(base) = basis.get(&pivot) {
                for (&row, &entry) in base {
                    let z = add_mod(*vector.get(&row).unwrap_or(&0), P - mul_mod(value, entry));
                    if z == 0 { vector.remove(&row); } else { vector.insert(row, z); }
                }
            } else {
                let inverse = inv(value);
                for entry in vector.values_mut() { *entry = mul_mod(*entry, inverse); }
                basis.insert(pivot, vector);
                break;
            }
        }
    }
    basis.len()
}

fn component_rows(cutoff: usize) -> Vec<Row> {
    let mut rows = Vec::new();
    for (component, frame_character) in [(0, -1), (1, 1), (2, 1)] {
        for ud in 0..=1 {
            for total in 0..=cutoff {
                for ad in 0..=total {
                    let coefficient_character = if ad % 2 == 0 { 1 } else { -1 };
                    if coefficient_character * frame_character == 1 {
                        rows.push((component, ud, ad, total - ad));
                    }
                }
            }
        }
    }
    rows
}

fn emit(parts: &[Poly; 3], cutoff: usize, position: &HashMap<Row, usize>, columns: &mut Vec<Column>) {
    let nonempty = parts.iter().any(|part| !part.is_empty());
    if !nonempty { return; }
    // Whole-column admission: inspect every component before plus projection.
    if parts.iter().flat_map(|part| part.keys()).any(|&(ud, ad, bd)| ud >= 2 || ad + bd > cutoff) {
        return;
    }
    let mut column = Column::new();
    for (component, part) in parts.iter().enumerate() {
        for (&(ud, ad, bd), &coefficient) in part {
            if let Some(&row) = position.get(&(component, ud, ad, bd)) {
                let z = add_mod(*column.get(&row).unwrap_or(&0), coefficient);
                if z == 0 { column.remove(&row); } else { column.insert(row, z); }
            }
        }
    }
    if !column.is_empty() { columns.push(column); }
}

fn census(cutoff: usize) -> (usize, usize, usize) {
    let one = mon(0, 0, 0, 1);
    let u = mon(1, 0, 0, 1);
    let a = mon(0, 1, 0, 1);
    let b = mon(0, 0, 1, 1);
    let l1 = add(&[b.clone(), one, scale(&u, P - 1)]);
    let l2_minus = add(&[a.clone(), scale(&u, P - inv(2))]);
    let l2_plus = add(&[a.clone(), scale(&u, inv(2))]);
    let euler = [scale(&a, inv(4)), Poly::new(), scale(&u, inv(2))];

    let rows = component_rows(cutoff);
    let position: HashMap<_, _> = rows.iter().enumerate().map(|(i, &row)| (row, i)).collect();
    let mut columns = Vec::new();

    for (sa, sb) in [(1usize, 1usize), (1, 0), (0, 1), (0, 0)] {
        let (ea, eb) = (2 - sa, 2 - sb);
        for l2 in [&l2_minus, &l2_plus] {
            let base = mul(&power(&l1, ea), &power(l2, eb));
            for total in 0..=cutoff {
                for ad in 0..=total {
                    let f = mon(0, ad, total - ad, 1);
                    let m = mul(&f, &base);

                    let mut cp = scale(&mul(&derivative(&f, 2), &base), P - 1);
                    if sa != 0 {
                        cp = add(&[cp, scale(&mul(&mul(&f, &power(&l1, ea - 1)), &power(l2, eb)), sa as i64)]);
                    }
                    let hp = [Poly::new(), scale(&m, mul_mod(3, inv(2))), Poly::new()];
                    let pparts = std::array::from_fn(|i| add(&[hp[i].clone(), mul(&cp, &euler[i])]));
                    emit(&pparts, cutoff, &position, &mut columns);

                    let mut cq = mul(&derivative(&f, 1), &base);
                    if sb != 0 {
                        cq = add(&[cq, scale(&mul(&mul(&f, &power(&l1, ea)), &power(l2, eb - 1)), P - sb as i64)]);
                    }
                    let hq = [scale(&m, P - mul_mod(3, inv(2))), Poly::new(), Poly::new()];
                    let qparts = std::array::from_fn(|i| add(&[hq[i].clone(), mul(&cq, &euler[i])]));
                    emit(&qparts, cutoff, &position, &mut columns);
                }
            }
        }
    }

    for total in 0..=cutoff {
        for ad in 0..=total {
            let p = mon(0, ad, total - ad, 1);
            let parts = std::array::from_fn(|i| mul(&p, &euler[i]));
            emit(&parts, cutoff, &position, &mut columns);
        }
    }

    let full_dimension = rows.len() - rank(columns.clone());
    let zero_rows: Vec<_> = rows.iter().copied().filter(|row| row.1 == 0).collect();
    let zero_position: HashMap<_, _> = zero_rows.iter().enumerate().map(|(i, &row)| (row, i)).collect();
    let zero_columns = columns.into_iter().filter_map(|column| {
        let restricted: Column = column.into_iter().filter_map(|(row, value)| {
            (rows[row].1 == 0).then(|| (zero_position[&rows[row]], value))
        }).collect();
        (!restricted.is_empty()).then_some(restricted)
    }).collect();
    let special_dimension = zero_rows.len() - rank(zero_columns);
    let torsion_dimension = 2 * special_dimension - full_dimension;
    (full_dimension, special_dimension, torsion_dimension)
}

fn main() {
    let expected = [(12, (105, 68, 31)), (16, (155, 106, 57)), (20, (213, 152, 91)), (24, (279, 206, 133))];
    for (cutoff, wanted) in expected {
        let result = census(cutoff);
        assert_eq!(result, wanted);
        println!("D={cutoff}: dim_Q={} dim_Q_mod_u={} t={}", result.0, result.1, result.2);
    }
    println!("{{\"schema\":\"marici.benincasa.q_d_census.v1\",\"closed_form\":\"t_D=(D/2)^2-D/2+1\"}}");
}
